src/orbital.rs

```diff
@@ -42,50 +55,79 @@ impl Drop for OrbitalMap {
+        Ok(Self {
+            handle,
+            width: 0,
+            height: 0,
+        })
+    }
+
+    pub fn resize(&mut self, width: u32, height: u32) -> Result<(), SoftBufferError> {
+        self.width = width;
+        self.height = height;
+        Ok(())
     }
 
-    pub(crate) unsafe fn set_buffer(&mut self, buffer: &[u32], width_u16: u16, height_u16: u16) {
-        let window_fd = self.handle.window as usize;
+    fn window_fd(&self) -> usize {
+        self.handle.window as usize
+    }
 
-        // Read the current width and size
+    // Read the current width and size
+    fn window_size(&self) -> (usize, usize) {
         let mut window_width = 0;
         let mut window_height = 0;
-        {
-            let mut buf: [u8; 4096] = [0; 4096];
-            let count = syscall::fpath(window_fd, &mut buf).unwrap();
-            let path = str::from_utf8(&buf[..count]).unwrap();
-            // orbital:/x/y/w/h/t
-            let mut parts = path.split('/').skip(3);
-            if let Some(w) = parts.next() {
-                window_width = w.parse::<usize>().unwrap_or(0);
-            }
-            if let Some(h) = parts.next() {
-                window_height = h.parse::<usize>().unwrap_or(0);
-            }
+
+        let mut buf: [u8; 4096] = [0; 4096];
+        let count = syscall::fpath(self.window_fd(), &mut buf).unwrap();
+        let path = str::from_utf8(&buf[..count]).unwrap();
+        // orbital:/x/y/w/h/t
+        let mut parts = path.split('/').skip(3);
+        if let Some(w) = parts.next() {
+            window_width = w.parse::<usize>().unwrap_or(0);
+        }
+        if let Some(h) = parts.next() {
+            window_height = h.parse::<usize>().unwrap_or(0);
         }
 
+        (window_width, window_height)
+    }
+
+    pub fn buffer_mut(&mut self) -> Result<BufferImpl, SoftBufferError> {
+        let (window_width, window_height) = self.window_size();
+        let pixels = if self.width as usize == window_width && self.height as usize == window_height
+        {
+            Pixels::Mapping(
+                unsafe { OrbitalMap::new(self.window_fd(), window_width * window_height * 4) }
+                    .expect("failed to map orbital window"),
+            )
+        } else {
+            Pixels::Buffer(vec![0; self.width as usize * self.height as usize])
+        };
+        Ok(BufferImpl { imp: self, pixels })
+    }
+
+    fn set_buffer(&self, buffer: &[u32], width_u32: u32, height_u32: u32) {
+        // Read the current width and size
+        let (window_width, window_height) = self.window_size();
+
         {
             // Map window buffer
             let window_map =
-                unsafe { OrbitalMap::new(window_fd, window_width * window_height * 4) }
+                unsafe { OrbitalMap::new(self.window_fd(), window_width * window_height * 4) }
                     .expect("failed to map orbital window");
 
             // Window buffer is u32 color data in 0xAABBGGRR format
```

Comment:
@jackpot51 sorry for my naivety here, but this comment suggests that the buffer taken and returned is partly reversed compared to what is expected by the rest of the library. Isn't that a problem?

Reply:
Hm, good point. In my testing it *seemed* like the colors where the same on all platforms, but something isn't right with the documentation here, at least.

Reply:
In any case the existing implementation isn't swapping components and this comment is already here, so this isn't something introduced in this PR. (Though if the format didn't match that might impact the possibility of no-copy on Orbital.)